File: backend-services/user-service/app/schemas/employee.py
from datetime import date, datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field


class EmployeeBase(BaseModel):
    name: str = Field(..., min_length=1, max_length=255)
    designation: str = Field(..., min_length=1, max_length=255)
    phone_number: Optional[str] = Field(default=None, max_length=32)
    monthly_salary: Decimal = Field(..., ge=0)
    per_day_salary: Optional[Decimal] = Field(default=None, ge=0)
    regular_employee: bool = True
    daily_allowance: Optional[Decimal] = Field(default=None, ge=0)
    advance_payment_received: Optional[Decimal] = Field(default=None)
    advance_pending: Optional[Decimal] = Field(default=None)
    joining_date: Optional[date] = None


class EmployeeCreate(EmployeeBase):
    pass


class EmployeeUpdate(BaseModel):
    name: Optional[str] = Field(default=None, min_length=1, max_length=255)
    designation: Optional[str] = Field(default=None, min_length=1, max_length=255)
    phone_number: Optional[str] = Field(default=None, max_length=32)
    monthly_salary: Optional[Decimal] = Field(default=None, ge=0)
    per_day_salary: Optional[Decimal] = Field(default=None, ge=0)
    regular_employee: Optional[bool] = None
    daily_allowance: Optional[Decimal] = Field(default=None, ge=0)
    advance_payment_received: Optional[Decimal] = Field(default=None)
    advance_pending: Optional[Decimal] = Field(default=None)
    joining_date: Optional[date] = None


class EmployeeResponse(EmployeeBase):
    id: int
    employee_code: str
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class PaginatedEmployees(BaseModel):
    items: list[EmployeeResponse]
    total: int
    page: int
    page_size: int
    pages: int


class EmployeeMonthlySummary(BaseModel):
    employee: EmployeeResponse
    month: str
    start_date: date
    end_date: date
    present_days: int
    full_days: int
    half_days: int
    penalty_fridays: int
    penalty_friday_dates: list[date]
    overtime_full_days: int
    overtime_total_minutes: int
    overtime_total_hours: float
    friday_bonus_days: Decimal
    paid_leave_days: int
    unpaid_leave_days: int
    total_worked_minutes: int
    total_worked_hours: float
    total_advances: Decimal
    initial_advance: Decimal | None
    total_month_advances: Decimal
    base_pay_for_period: Decimal
    pending_advances_balance: Decimal
    monthly_salary: Decimal
    per_day_salary_for_month: Decimal
    net_payable: Decimal
    penalty_deduction_amount: Decimal
    overtime_credit_amount: Decimal
    salary_paid: bool = False
    salary_paid_on: Optional[datetime] = None
    salary_paid_amount: Optional[Decimal] = None
    salary_payment_id: Optional[int] = None
    salary_paid_note: Optional[str] = None
    salary_paid_additional_advance: Optional[Decimal] = None
    salary_paid_advance_reduction: Optional[Decimal] = None
    salary_remaining_initial_advance: Optional[Decimal] = None
    salary_pending_advance_after: Optional[Decimal] = None


class PaginatedEmployeeMonthlySummaries(BaseModel):
    items: list[EmployeeMonthlySummary]
    total: int
    page: int
    page_size: int
    pages: int

File: admin-frontend/src/types/employee.ts
export type Employee = {
  id: number;
  employee_code: string;
  name: string;
  designation: string;
  phone_number: string | null;
  monthly_salary: string;
  per_day_salary: string | null;
  regular_employee: boolean;
  daily_allowance: string | null;
  advance_payment_received: string | null;
  advance_pending: string | null;
  joining_date: string | null;
  created_at: string;
  updated_at: string;
};

export type PaginatedEmployees = {
  items: Employee[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type AttendanceRecord = {
  id: number;
  employee_id: number;
  work_date: string;
  clock_in: string;
  clock_out: string | null;
  worked_minutes: number;
  day_type: "pending" | "half" | "full" | string;
};

export type DailyAdvance = {
  id: number;
  employee_id: number;
  amount: string;
  note: string | null;
  recorded_for: string;
  created_at: string;
};

export type PaginatedAttendance = {
  items: AttendanceRecord[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type PaginatedDailyAdvance = {
  items: DailyAdvance[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type EmployeeMonthSummary = {
  employee: Employee;
  month: string;
  start_date: string;
  end_date: string;
  present_days: number;
  full_days: number;
  half_days: number;
  paid_leave_days: number;
  unpaid_leave_days: number;
  penalty_fridays: number;
  penalty_friday_dates: string[];
  overtime_full_days: number;
  overtime_total_minutes: number;
  overtime_total_hours: number;
  friday_bonus_days: string;
  total_worked_minutes: number;
  total_worked_hours: number;
  total_advances: string;
  initial_advance: string | null;
  total_month_advances: string;
  base_pay_for_period: string;
  pending_advances_balance: string;
  per_day_salary_for_month: string;
  monthly_salary: string;
  net_payable: string;
  penalty_deduction_amount: string;
  overtime_credit_amount: string;
  salary_paid: boolean;
  salary_paid_on: string | null;
  salary_paid_amount: string | null;
  salary_payment_id: number | null;
  salary_paid_note: string | null;
  salary_paid_additional_advance: string | null;
  salary_paid_advance_reduction: string | null;
  salary_remaining_initial_advance: string | null;
  salary_pending_advance_after: string | null;
};

export type LeaveRecord = {
  id: number;
  employee_id: number;
  start_date: string;
  end_date: string;
  leave_type: "paid" | "unpaid";
  reason: string | null;
  created_at: string;
  is_penalty?: boolean;
};

export type PaginatedLeaveRecords = {
  items: LeaveRecord[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type PaginatedPayroll = {
  items: EmployeeMonthSummary[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};

export type SalaryPayment = {
  id: number;
  employee_id: number;
  month: string;
  period_start: string;
  period_end: string;
  paid_amount: string;
  additional_advance: string;
  advance_reduction_amount: string;
  initial_advance_after: string;
  carry_forward_advance: string;
  pending_advance_after: string;
  status: string;
  note: string | null;
  paid_on: string;
  created_at: string;
};

export type PaginatedSalaryPayments = {
  items: SalaryPayment[];
  total: number;
  page: number;
  page_size: number;
  pages: number;
};
